compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -387,8 +391,91 @@ fn link_rlib<'a, B: ArchiveBuilder<'a>>(
+    // platform equivalent) to include all object files from an rlib into the
+    // final dylib itself. This causes linkers to iterate and try to include all
+    // files located in an archive, so if metadata is stored in an archive then
+    // it needs to be of a form that the linker will be able to process.
+    //
+    // Note, though, that we don't actually want this metadata to show up in any
+    // final output of the compiler. Instead this is purely for rustc's own
+    // metadata tracking purposes.
+    //
+    // With the above in mind, each "flavor" of object format gets special
+    // handling here depending on the target:
+    //
+    // * MachO - macos-like targets will insert the metadata into a section that
+    //   is sort of fake dwarf debug info. Inspecting the source of the macos
+    //   linker this causes these sections to be skipped automatically because
+    //   it's not in an allowlist of otherwise well known dwarf section names to
+    //   go into the final artifact.
+    //
+    // * WebAssembly - we actually don't have any container format for this
+    //   target. WebAssembly doesn't support the `dylib` crate type anyway so
+    //   there's no need for us to support this at this time. Consequently the
+    //   metadata bytes are simply stored as-is into an rlib.
+    //
+    // * COFF - Windows-like targets create an object with a section that has
+    //   the `IMAGE_SCN_LNK_REMOVE` flag set which ensures that if the linker
+    //   ever sees the section it doesn't process it and it's removed.
+    //
+    // * ELF - All other targets are similar to Windows in that there's a
+    //   `SHF_EXCLUDE` flag we can set on sections in an object file to get
+    //   automatically removed from the final output.
+    //
+    // Note that this metdata format is kept in sync with
+    // `rustc_codegen_llvm/src/metadata.rs`.
+    fn create_metadata_object(sess: &Session, metadata: &[u8]) -> Vec<u8> {
+        let endianness = match sess.target.options.endian {
+            Endian::Little => Endianness::Little,
+            Endian::Big => Endianness::Big,
+        };
+        let architecture = match &sess.target.arch[..] {
+            "arm" => Architecture::Arm,
+            "aarch64" => Architecture::Aarch64,
+            "x86" => Architecture::I386,
+            "s390x" => Architecture::S390x,
+            "mips" => Architecture::Mips,
+            "x86_64" => Architecture::X86_64,
+
+            // This is used to handle all "other" targets. This includes targets
+            // like WebAssembly and those otherwise just not having support here
+            // at this time. For these targets the metadata file in the rlib is
+            // simply the raw metadata itself. This would break dylib
+            // compliation if it were supported, but for most of these targets
+            // dylib compilation is not supported anyway (such as WebAssembly).
```

Comment:
Riscv, sparc and powerpc do support dylibs.

Reply:
Yeah, it seems like a better solution is necessary here. I would perhaps intuitively reach for `object_format: ...` field in the target definition, but perhaps excluding just wasm here is sufficient (majority of the other targets, including the embedded ones, produce elf)

Reply:
It's not that the code here claims it doesn't, nor that I intentionally ignored to add support. The `object` crate currently only supports a limited set of architectures (listed here) and support would need to be added for more architectures. If anything here is tier 2 and fails to build then I can try to find the time to send PRs to `object` myself, otherwise I was hoping others could do it.

Reply:
Makes sense. Can the comment above be adjusted to specify that lack of support here refers specifically to lack in the `object` crate, rather than, say, the target toolchain?